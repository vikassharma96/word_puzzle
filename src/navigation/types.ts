import type {NativeStackScreenProps} from '@react-navigation/native-stack';

export type RootStackParamList = {
  Home: undefined;
  Puzzle: {categoryId: string};
  Score: undefined;
  LeadersBoard: undefined;
};

export type HomeScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Home'
>;
export type PuzzleScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Puzzle'
>;
export type ScoreScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Score'
>;
export type LeadersBoardScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'LeadersBoard'
>;
