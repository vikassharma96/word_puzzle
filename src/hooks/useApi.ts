import {useState} from 'react';

const useApi = (apiFunc: any) => {
  const [data, setData] = useState([]);
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);

  const request = async (...args: any[]) => {
    try {
      setData([]);
      setError(false);
      setLoading(true);
      const response = await apiFunc(...args);
      const jsonResponse = await response.json();
      setError(!response.ok);
      setData(jsonResponse);
      return jsonResponse;
    } catch (err: any) {
      setError(true);
    } finally {
      setLoading(false);
    }
  };

  return {data, error, loading, request};
};

export default useApi;
