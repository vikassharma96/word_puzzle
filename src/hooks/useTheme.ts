import {useContext} from 'react';
import Colors from '../styles/colors';
import {ThemeContext} from '../contexts/theme';

const useTheme = () => {
  const {theme} = useContext(ThemeContext);
  return Colors[theme];
};

const useThemeColor = (
  colorName: keyof typeof Colors.light & keyof typeof Colors.dark,
) => {
  return useTheme()[colorName];
};

export {useTheme, useThemeColor};
