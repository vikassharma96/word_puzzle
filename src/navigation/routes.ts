export enum Routes {
  HOME = 'Home',
  PUZZLE = 'Puzzle',
  SCORE = 'Score',
  LEADERS_BOARD = 'LeadersBoard',
}
