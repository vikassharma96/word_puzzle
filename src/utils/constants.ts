// Home screen constants
export const WORDS_PUZZLE = 'Words Puzzle';
export const API_ERROR_MSG = "Couldn't retrieve the categories.";
export const START = 'Start';
export const LEADERS_BOARD = 'Leaders Board';
export const ERROR_MSG = 'Please select category first';

// Puzzle screen constants
export const WRONG_TEXT = 'Sorry! \n Bad luck';
export const CORRECT_TEXT = 'Correct! \n Congratulations';
export const EARN_TEXT = 'You earn';
export const POINTS = 'Points';
export const SCORE_SUMMARY = 'Score Summay';
export const TOTAL_SCORE = 'Your total score is';
export const SHARE = 'Share';

// common constants
export const ERROR = 'Error!';
export const RETRY = 'Retry';
export const BUTTON_TEXT_OK = 'Ok';
export const BUTTON_TEXT_CANCEL = 'Cancel';
export const SKIP = 'Skip';
export const NEXT = 'Next';
export const NO_DATA_AVAILABLE = 'No data available!';
export const WORD_PUZZLE_TEXT = 'Word Puzzle!';
export const SOMETHING_WENT_WRONG = 'Something went wrong!';
export const EXIT = 'Exit';
