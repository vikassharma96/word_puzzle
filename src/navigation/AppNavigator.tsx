import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import {Routes} from './routes';
import HomeScreen from '../views/Home';
import {RootStackParamList} from './types';
import PuzzleScreen from '../views/Puzzle';
import LeadersBoardScreen from '../views/LeadersBoard';
import Screen from '../components/Screen';

const Stack = createNativeStackNavigator<RootStackParamList>();

const AppNavigator = () => {
  return (
    <NavigationContainer>
      <Screen>
        <Stack.Navigator>
          <Stack.Screen name={Routes.HOME} component={HomeScreen} />
          <Stack.Screen name={Routes.PUZZLE} component={PuzzleScreen} />
          <Stack.Screen
            name={Routes.LEADERS_BOARD}
            component={LeadersBoardScreen}
          />
        </Stack.Navigator>
      </Screen>
    </NavigationContainer>
  );
};

export default AppNavigator;
