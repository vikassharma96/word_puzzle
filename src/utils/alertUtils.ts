import {Alert} from 'react-native';
import {BUTTON_TEXT_CANCEL, BUTTON_TEXT_OK} from './constants';

type AlertFunction = {
  title: string;
  message: string;
  singleButton?: boolean;
  positiveText?: string;
  negativeText?: string;
  cancelable?: boolean;
  positiveAction?: () => void;
  negativeAction?: () => void;
  positiveActionStyle?: any;
  negativeActionStyle?: any;
};

export const showAlert = ({
  title,
  message,
  singleButton = true,
  positiveText = BUTTON_TEXT_OK,
  negativeText = BUTTON_TEXT_CANCEL,
  cancelable = false,
  positiveAction,
  negativeAction,
  positiveActionStyle,
  negativeActionStyle,
}: AlertFunction) => {
  const actionButtons = [
    {text: positiveText, onPress: positiveAction, style: positiveActionStyle},
  ];
  if (!singleButton) {
    actionButtons.unshift({
      text: negativeText,
      onPress: negativeAction,
      style: negativeActionStyle,
    });
  }
  Alert.alert(title, message, actionButtons, {
    cancelable: cancelable,
  });
};
