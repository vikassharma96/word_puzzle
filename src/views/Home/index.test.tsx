import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import HomeScreen from '.';
import {Routes} from '../../navigation/routes';
import {LEADERS_BOARD} from '../../utils/constants';
import * as hooks from '../../hooks/useApi';

describe('views/Home', () => {
  const mockNavigation = {
    navigate: jest.fn(),
  };
  const mockCategories = () => [
    {
      id: '8f6aba43-f445-4449',
      name: 'Cities',
    },
    {
      id: '01f4185b-18e6-4ed5',
      name: 'Food',
    },
  ];
  const component = <HomeScreen navigation={mockNavigation} />;
  it('should match snapshots', () => {
    const componentTree = renderer.create(component).toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  it('should be able to navigate to leadersboard screen', () => {
    const {getByText} = render(component);
    const leadersBoardBtn = getByText(LEADERS_BOARD);
    expect(leadersBoardBtn).toBeDefined();
    fireEvent.press(leadersBoardBtn);
    expect(mockNavigation.navigate).toBeCalledWith(Routes.LEADERS_BOARD);
  });

  it.skip('shouble be able to render categories', () => {
    jest.spyOn(hooks, 'useApi').mockImplementation(() => ({
      data: [],
      error: false,
      loading: false,
      request: mockCategories,
    }));
    const {getByText} = render(component);
    expect(getByText('Cities')).toBeDefined();
  });
});
