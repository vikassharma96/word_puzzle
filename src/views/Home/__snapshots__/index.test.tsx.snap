// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/Home should match snapshots 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "borderRadius": 6,
          "borderWidth": 1.5,
          "marginHorizontal": 8,
          "paddingHorizontal": 16,
          "paddingVertical": 6,
          "textAlign": "center",
        },
        {
          "backgroundColor": "#CEDBEB",
        },
        {
          "alignSelf": "center",
          "marginTop": "4%",
          "textAlign": "center",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Medium",
            "fontSize": 14,
          },
          {
            "color": "#000",
          },
          undefined,
        ]
      }
    >
      Words Puzzle
    </Text>
  </View>
  <View
    style={
      {
        "backgroundColor": "#f8f8f8",
        "flex": 0.8,
        "marginTop": "4%",
        "paddingTop": "2%",
      }
    }
  >
    <RCTScrollView
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View />
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "flex": 0.2,
        "paddingTop": "3%",
        "zIndex": -1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#F0CFCD",
          "borderRadius": 6,
          "borderWidth": 1.5,
          "marginHorizontal": 8,
          "minWidth": "50%",
          "opacity": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 6,
          "textAlign": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Poppins-Medium",
              "fontSize": 14,
            },
            {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Start
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "#fff",
          "borderRadius": 6,
          "borderWidth": 0,
          "marginHorizontal": 8,
          "marginTop": "2%",
          "opacity": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 6,
          "textAlign": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Poppins-Medium",
              "fontSize": 14,
            },
            {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Leaders Board
      </Text>
    </View>
  </View>
</View>
`;
